packages/vite/src/node/server/middlewares/htmlFallback.ts

```diff
@@ -25,6 +26,13 @@ export function htmlFallbackMiddleware(
           return spaFallback ? `/index.html` : request.url
         },
       },
+      // don't rewrite paths ending with .html
+      {
+        from: /\.html$/,
+        to({ request }: any) {
+          return request.url
+        },
+      },
```

Comment:
Because `serveStaticMiddleware` doesn't handle html files and the fallback middleware runs before html middleware.
https://github.com/vitejs/vite/blob/bd9b749fe5c7d390a48368ca3a2f4ac55e730607/packages/vite/src/node/server/middlewares/static.ts#L91-L102